### PR TITLE
Fix BottomSheet navigation hardware back button support

### PR DESCRIPTION
**Pull request by dcalhoun**
## Description
An earlier addition of `setHeight` to the dependencies array (https://github.com/WordPress/gutenberg/commit/2a8d738b40a5f803045eda68e29eca8c08918af0) meant the callback was recreated multiple times, which led to unintentional updates to the hardware button handler. This resulted in navigating backwards multiple screen instead of one when pressing the hardward back button in a deeply nested screen.

`onHandleHardwareButtonPress` stores a single value, which means future invocations from sibling screens can replace the callback for the currently active screen. Currently, the empty dependency array passed to `useCallback` here is what prevents erroneous callback replacements, but leveraging memoization to achieve this is brittle and explicitly discouraged in the React documentation. https://reactjs.org/docs/hooks-reference.html#usememo

Ideally, we refactor `onHandleHardwareButtonPress` to manage multiple callbacks triggered based upon which screen is currently active. I opted to postpone that work to avoid delaying a merge of the targeted RN upgrade PR. 

Related: https://github.com/WordPress/gutenberg/pull/36328#discussion_r768897546

## How has this been tested?
1. Open the editor on an Android device. 
1. Add a Buttons block. 
2. Open the Buttons block settings. 
3. Tap _Background color_. 
4. Scroll the palette colors left, and tap _Custom_. 
5. Tap the hardware back button. 
6. ℹ️ **Expected:** The palettes are displayed. 
5. Tap the hardware back button. 
6. ℹ️ **Expected:** The Buttons block settings are displayed. 
5. Tap the hardware back button. 
6. ℹ️ **Expected:** The bottom sheet is closed. 
5. Tap the hardware back button. 
6. ℹ️ **Expected:** The app is closed. 

## Screenshots 

https://user-images.githubusercontent.com/438664/146266501-e94c2e4c-a28e-44a4-b053-0489d42afff4.mp4

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by dcalhoun** [comment]
I believe the failing Admin e2e CI tasks to be unrelated and likely a result from this branch being outdated with `trunk`. However, I would prefer to update the target branch with `trunk` rather than this branch. I.e. I think it is safe to ignore the failure for this PR specifically.

**Review by hypest — APPROVED** [review]
LGTM!

I've verified that this fixes the missing back-button step in the nav history 🎉

That said, I have to say that I still find it hard to follow what the code does, in terms of the flow of callbacks for navigation in the bottom sheet. That's not an "issue" introduced by this PR though so, not a blocker.
